### PR TITLE
Implement hardware connectivity confirmation screens

### PR DESCRIPTION
**Pull request by darkwing**
Designs found here:  https://www.figma.com/file/6Px3A6vVFGy1hYbEUXi343/Hardware-Wallets?node-id=2%3A2020

## ToDo:

- [x] Implement form fields in popover
- [x] Implement popover padding and other styling
- [x] Remove hardcoded debug values
- [x] Implement polling of connectivity status
- [x] Remove LedgerInstructionsField
- [x] Update Ledger keyring library

## Manual Testing
1.  Pull down this PR
2. In `node_modules/@metamask/eth-ledger-bridge-keyring/index.js`, change the `BRIDGE_URL` to `'https://darkwing.github.io/eth-ledger-bridge-keyring'`; that address includes the `gh-pages` portion of this project (https://github.com/MetaMask/eth-ledger-bridge-keyring/pull/147).  Also, comment out the following check for origin:

```
/*
      if (origin !== this._getOrigin()) {
        return false
      }
*/
```

3. Connect a Ledger address, then try to do a basic send on testnet -- you will see the connectivity messaging and functionality!

## Considerations
1.  Dropdown mode vs. full screen mode
2. The different Ledger connectivity options

### CODE REVIEW

**Comment by jpuri** [comment]
Hey @darkwing : seems there is no activity on the PR from sometime.

**Comment by darkwing** [comment]
@jpuri This PR is basically done; the majority of the work needs to be done in the `eth-ledger-bridge-keyring` repo.

File: ui/pages/confirm-transaction-base/hardware-connectivity/hardware-connectivity-content.js
Comment (darkwing) [inline ui/pages/confirm-transaction-base/hardware-connectivity/hardware-connectivity-content.js:38]: ToDo:  Use localization for those steps.

File: ui/pages/confirm-transaction-base/hardware-connectivity/hardware-connectivity-content.js
Comment (darkwing) [inline ui/pages/confirm-transaction-base/hardware-connectivity/hardware-connectivity-content.js:33]: TODO:  Add `alt=""` here.

File: ui/pages/confirm-transaction-base/confirm-transaction-base.component.js
Comment (darkwing) [inline ui/pages/confirm-transaction-base/confirm-transaction-base.component.js:196]: TODO: Change `2000` to a constant.

File: ui/pages/confirm-transaction-base/hardware-connectivity/hardware-connectivity-content.js
Comment (KOOLMOLL) [reply ui/pages/confirm-transaction-base/hardware-connectivity/hardware-connectivity-content.js]: ToDo: Use localization for those steps.

File: ui/components/app/confirm-page-container/confirm-page-container-content/confirm-page-container-content.component.js
Comment (danjm) [inline ui/components/app/confirm-page-container/confirm-page-container-content/confirm-page-container-content.component.js:132]: Might be good to generalize this variable name. Within this component, this variable just determines whether we should show `ConfirmPageContainerSummary` or not. So we could change this to `hideConfirmPageContainerSummary`

File: ui/pages/confirm-transaction-base/confirm-transaction-base.component.js
Comment (danjm) [inline ui/pages/confirm-transaction-base/confirm-transaction-base.component.js:1079]: Should we also clear the interval on a window `beforeunload` event?

**Review by danjm — CHANGES_REQUESTED** [review]
This looks pretty good! Left a couple of small comments. Should we also get a design review from @holantonela ?

File: ui/components/app/confirm-page-container/confirm-page-container-content/confirm-page-container-content.component.js
Comment (darkwing) [reply ui/components/app/confirm-page-container/confirm-page-container-content/confirm-page-container-content.component.js:132]: I prefer not to name things with a negative word; also, `showingHardwareConnectionContents` is akin to `showBuyModal`, representing whether or not we are showing the popover.  Maybe you'd prefer `showingHardwareConnectionPopover`?

Comment (danjm) [reply ui/components/app/confirm-page-container/confirm-page-container-content/confirm-page-container-content.component.js:132]: I was mostly thinking to remove the reference to the hardware wallet. In the context of this file, the hardware wallet is irrelevant, the boolean property just indicates whether or not to show `ConfirmPageContainerSummary`... the parent of this component can decide to hide or show that for whatever reason it chooses (not just because of the hardware wallet being connected)

**Review by danjm — CHANGES_REQUESTED** [review]
This looks good to me. We just need to get the ledger keyring library updated

**Comment by segun** [comment]
On this branch, I see ledger is not connected even though my device is connected

**Comment by AlexJupiter** [comment]
@holantonela @darkwing I've created a metrics ticket for this experience here: https://github.com/MetaMask/metamask-extension/issues/14014

File: ui/components/ui/actionable-message/index.scss
Comment (darkwing) [inline ui/components/ui/actionable-message/index.scss:122]: TODO: Update these colors

File: ui/pages/create-account/connect-hardware/connect-hardware-path.js
Comment (darkwing) [inline ui/pages/create-account/connect-hardware/connect-hardware-path.js:24]: Add `isRequired` here.

File: ui/pages/settings/advanced-tab/ledger-transport-dropdown.js
Comment (digiwand) [inline ui/pages/settings/advanced-tab/ledger-transport-dropdown.js:87]: I see that this id name was copied over verbatim, but might I suggest we change this to match the file name? This is to promote more robust code standards and avoid naming collisions.

**Review by digiwand — COMMENTED** [review]
Took a wing at reviewing this code though it's my first time looking at the hardware logic. Left some suggestions and saw no blockers! 🎉 

It's also my first time running the development code with the ledger connectivity. I checked out this branch and successfully sent a transaction with this code. I'm going to step through some of the code in action before submitting a final approval

File: ui/pages/confirm-transaction-base/hardware-connectivity/hardware-connectivity-content.js
Comment (digiwand) [inline ui/pages/confirm-transaction-base/hardware-connectivity/hardware-connectivity-content.js:31]: Q: Will more tests and storybook logic be added to support the new changes?

File: ui/pages/confirm-transaction-base/confirm-transaction-base.component.js
Comment (darkwing) [reply ui/pages/confirm-transaction-base/confirm-transaction-base.component.js:181]: At one point we required a return value or linting would fail, but maybe that's now gone.

File: ui/pages/confirm-transaction-base/confirm-transaction-base.component.js
Comment (darkwing) [reply ui/pages/confirm-transaction-base/confirm-transaction-base.component.js:184]: Yes -- if something is somehow faulty with our code, we don't want to prevent users from being able to do whatever they'd like.

File: ui/pages/confirm-transaction-base/hardware-connectivity/hardware-connectivity-message.js
Comment (darkwing) [reply ui/pages/confirm-transaction-base/hardware-connectivity/hardware-connectivity-message.js:10]: Good shout!

File: ui/pages/confirm-transaction-base/confirm-transaction-base.component.js
Comment (darkwing) [reply ui/pages/confirm-transaction-base/confirm-transaction-base.component.js:651]: Good shout!

File: ui/pages/settings/advanced-tab/ledger-transport-dropdown.js
Comment (darkwing) [reply ui/pages/settings/advanced-tab/ledger-transport-dropdown.js:87]: Cool with me!

File: ui/pages/confirm-transaction-base/hardware-connectivity/hardware-connectivity-content.js
Comment (darkwing) [reply ui/pages/confirm-transaction-base/hardware-connectivity/hardware-connectivity-content.js:24]: Yes!!!!

**Review by digiwand — COMMENTED** [review]
a couple of more comments coming in after stepping through the code and some more inspection

File: ui/pages/create-account/connect-hardware/connect-hardware-path.js
Comment (darkwing) [reply ui/pages/create-account/connect-hardware/connect-hardware-path.js:1]: Oh, we changed direction after this was created, so not used anymore.  Really good find!

File: ui/pages/settings/advanced-tab/ledger-transport-dropdown.js
Comment (darkwing) [reply ui/pages/settings/advanced-tab/ledger-transport-dropdown.js:1]: Done!  Great catch!

**Review by brad-decker — DISMISSED** [review]
Tested with my ledger over the weekend using the library update and this is working very well.

File: ui/store/actions.js
Comment (danjm) [inline ui/store/actions.js:413]: `Looking for your ${capitalize(deviceName)}...` needs to be localized

**Review by danjm — DISMISSED** [review]
Only one comment from me. This looks good! I tested the solution for when webhid is disconnected, and it works.

File: ui/store/actions.js
Comment (darkwing) [reply ui/store/actions.js:413]: This isn't new -- it's an existing issue.  We don't have `this.context.t` available in actions.  I believe there's an issue for this open already.

**Review by brad-decker — DISMISSED** [review]

File: ui/pages/confirm-transaction-base/confirm-transaction-base.component.js
Comment (digiwand) [reply ui/pages/confirm-transaction-base/confirm-transaction-base.component.js:1129]: @darkwing bump - just want to double-check this won't cause any memory leaks. If we're unsure, we can add the remove listener just in case.

```suggestion
    window.removeEventListener('beforeunload', () => this._clearPollingInterval);
    this._clearPollingInterval();
```

I see `this._removeBeforeUnload();` is being called above, but it's removing the event with a different callback
```
  _removeBeforeUnload = () => {
    window.removeEventListener('beforeunload', this._beforeUnloadForGasPolling);
  };
```

**Review by digiwand — COMMENTED** [review]
LGTM and appears to be working well on Chrome on my macOS! I left some optional suggestions. 

The one thing holding me back from approving is this event listener that I believe we want to remove on unmount: 
https://github.com/MetaMask/metamask-extension/pull/12725#discussion_r870849004

**Comment by darkwing** [comment]
Addressed feedback!

**Review by digiwand — DISMISSED** [review]
LGTM! 🙌🏼

**Comment by danjm** [comment]
@darkwing This needs to be rebased onto develop to get all tests passing.

**Review by digiwand — APPROVED** [review]
🙌

**Comment by brad-decker** [comment]
@darkwing @danjm @hilvmason @kevinghim this seems like important work that we failed to get landed and is now seriously out of date. @darkwing i'm sorry that happened to you. Should we put this PR into sprint queue to prioritize review / rebase time for it?

**Comment by darkwing** [comment]
@brad-decker I don't think so.  The priority is MV3, and this PR could conflict with the work the team is doing on hardware wallet stuff for MV3.  Also, this PR and its dependency heavily rely on `setInterval` and `setTimeout`, which will no longer work with MV3, so it's best that the MV3 work gets done first.

**Comment by brad-decker** [comment]
@darkwing sounds good, but lets plan to revisit this conversation and have the review / rebase time dedicated as part of a future sprint after MV3. In the meantime, i'll convert to draft.

**Comment by holantonela** [comment]
We now have dedicated extension development capacity at the Key Management team, so perhaps @darkwing and @HowardBraham can work together on this after MV3 priors get updated. @AlexJupiter will coordinate it.

**Pull request closed without merging**
